src/procs/proc_verify.rs

```diff
@@ -1,9 +1,59 @@
 use crate::{
-    ast::{Block, Direction, ProcDecl, SpanVariant, Spanned, StmtKind},
+    ast::{
+        self, BinOpKind, Block, Direction, ExprData, ExprKind, ProcDecl, Shared, Span, SpanVariant,
+        Spanned, StmtKind, TyKind,
+    },
     driver::core_verify::CoreVerifyTask,
     slicing::{wrap_with_error_message, wrap_with_success_message},
 };
 
+/// Encode the preexpectations as a formula for which a model is a nontrivial bound
+/// For proc:
+///     goal: Find a model such that (pre1 ⊓ pre2 ⊓...) > 0
+/// For coproc:
+///     goal: Find a model such that (pre1 ⊔ pre2 ⊔...) < top
+pub fn encode_preexpectation(proc: &ProcDecl) -> Option<(Direction, Block)> {
+    let direction = proc.direction;
+
+    let body_ref = proc.body.borrow();
+    let body = match &*body_ref {
+        Some(body) => body,
+        None => return None,
+    };
+    let mut block = Spanned::new(body.span, vec![]);
+
+    fn get_combination(direction: Direction) -> ast::expr::BinOpKind {
+        let combination = match direction {
+            Direction::Down => BinOpKind::Inf,
+            Direction::Up => BinOpKind::Sup,
+        };
+        combination
+    }
+
+    // connect the preexpectations through the correct operator
+    let mut build_expr = proc
```

Comment:
Why not just use a [`reduce`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.reduce)? That seems much simpler and less error-prone